File: Pythonleetcode/Leetcode322.py
class Solution(object):
    def coinChange(self, coins, amount):
        dp=[0]*(amount+1)
        for i in range(1,amount+1):
            temp=[dp[i-k]+1 for k in coins if k<=i and dp[i-k]!=-1]
            if len(temp)>0:
                dp[i]=min(temp)
            else:
                dp[i]=-1
        # print(dp)
        return dp[-1]

print(Solution().coinChange(coins = [1,2,5], amount = 11))
print(Solution().coinChange(coins = [2], amount = 3))
print(Solution().coinChange(coins = [1], amount = 0))
print(Solution().coinChange(coins = [1], amount = 1))

File: CppPractice/CppLeetcode/Leetcode322.cpp
class Solution {
public:
    int coinChange(vector<int>& coins, int amount) {
        if(amount==0){return 0 ;}
        // sort(coins.begin(),coins.end());
        vector<int> dp(amount+1,-1);
        dp[0]=0;
        for(int i=1;i<dp.size();i++)
        {
            for(int coin:coins)
            {
                if(coin<=i && dp[i-coin]!=-1)
                {
                    if(dp[i]==-1){dp[i]=dp[i-coin]+1;}
                    else{dp[i] = min(dp[i-coin]+1,dp[i]);}
                }
            }
        }
        return dp[dp.size()-1];
    }
};